docs: add domain model documentation (#761)
eclipse-dataspaceconnector/DataSpaceConnector: docs/domain-model.md

# Domain Model
---

## Definitions:

### EDC connector

> <p style='text-align: justify;'> EDC is an interoperable, cross-organization framework for data sharing between 2 or more Organizations in a trusted and

Reviewer: The domain model file should not explain what the EDC is. This is duplicated information, and if not, should be placed elsewhere.

  facilitate billing or resolve a conflict. Furthermore, the Data Provider can use Data Apps to enrich or transform the
  data in some way, or to improve its quality (Data Apps are specific applications that can be integrated into the data
  exchange workflow between two or more participants in the Industrial Data Space)<sup>2</sup> The data owner can also be the data
  provider, when it is not, then a contract is needed between the client and provider in order to start a transfer of
  data. </p>

* **Data consumer**:

  <p style='text-align: justify;'>The Data Consumer receives data from a Data Provider. Before the connection to a Data Provider can be established, the
  Data Consumer can search for existing datasets by making an inquiry at a Broker Service Provider (provides the required
  metadata for the Data Consumer to connect to a Data Provider.). Alternatively (In case the information to connect with
  the Data Provider is already known to the Data Consumer), the Data Consumer can establish a connection with a Data
  Provider directly<sup>2</sup> The data user, can also be the data consumer.</p>

> ![img_3.png](dataConsumerDiagram.png)

Reviewer: I am not sure whether we should use an IDS diagram for explaining our domain model. Especially when this is about the integration of an EDC as IDS Connector in an ecosystem. The domain model document should explain our understanding of relations inside and outside the EDC. You could take a look at the [existing diagrams](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/tree/main/docs/diagrams).

Author: Hi @juliapampus, the code was already updated. Please check the newest version.

  policies, and an asset selector which links the contract to one or more assets.

* ### Contract offer

  The contract offer is a dynamic representation of the [`ContractDefinition`](#contract-definition)
  for a specific consumer and serves as protocol DTO for a particular contract negotiation. These contract offers are
  not persisted and will be regenerated on every request. The provider connector will only generate contract offers for
  those contract definitions where the inquiring organizations satisfy the policies established in the contract
  definitions access policy. A contract offer is always related to a single asset of the contract definition object (e.g
  for a `ContractDefinition` containing three `Asset` objects the provider will generate three `ContractOffer` objects )
  .

* ### Contract negotiation

  A `ContractNegotiation` captures the current state of the negotiation of a contract (`ContractOffer` ->
  `ContractAgreement`) between two parties. This process is inherently asynchronous, so the `ContractNegotiation`
  objects are stored in a backing data store (`ContractNegotiationStore`).

* ### Contract agreement

>   Represents the agreed-upon terms of usage of an asset between two parties including a start and an end date.

Reviewer: ```suggestion
  A contract agreement represents the agreed-upon terms of access and usage of an asset's data between two data space participants, including a start and an end date and further relevant information.
```

## Policy

Contract policies represent permitted and prohibited actions over a certain asset. These actions can be limited further
by constraints (temporal or spatial) and duties ("e.g. deletion of the data after 30 days"). Further information is
provided in a separate [section](Policies.md).

## Data request

A `DataRequest` is sent from the consumer to the provider after a successful contract negotiation to initiate the data
transfer. It references the [`Asset`](#asset) and [`Contract`](#contract-agreement) as well as information about
the [data destination](#data-address).

## Transfer process

> Analogous to the `ContractNegotiation` this object captures the current state of a data transfer. This process is

Reviewer: ```suggestion
Similar to the `ContractNegotiation`, this object captures the current state of a data transfer. This process is
```

  accessing the contract offers. In addition, the consumer must follow these policies when accessing the assets.

* ### Contract Offer

  The contract offer is a dynamic representation of the [contract definition](#contract-definition)
  for a specific consumer and will be used for the contract negotiation. This contract offers are not persistent and
  will be regenerated on every request. The provider EDC will only generate contract offers for the contract definitions
  where the inquiring organizations privileges satisfy the policies established in the contract definitions access
  policy. One contract offer is always related to only one asset of the contract definition (e. g. if the contract
  definition is related to 3 assets, the consumer will get 3 separate contract offers). In addition, the contract offer
  contains a policy which is the contract policy.

* ### Contract Negotiation

  Is the process when an organization, represented with a connector as a consumer, asks for access to a certain asset
  and supplies all the requirements specified in the contract offer. Some states of the contract negotiation are the

>   requesting of an asset, the offers asked in the negotiation, the approval or the declining of it.

Reviewer 2: _I_ know what you want to express, but this section is very misleading, and someone not familiar with the EDC code will get confused. 

Better:
_A `ContractNegotiation` captures the current state of the negotiation of a contract (`ContractOffer` -> `ContractAgreement`) between two parties. This process is inherently asynchronous, so the `ContractNegotiation` objects are stored in a backing data store (`ContractNegotiationStore`)._

> * ### Contract Agreement

Reviewer 2: ```suggestion
* ### Contract agreement
```

>   agreed usage between data provider and data consumer after a contract negotiation. It points to a contract offer and
>   can have expiry or cancellation date.

Reviewer 2: ```suggestion
  Represents the agreed-upon terms of usage of an asset between two parties including a start and an end date.
```

## Policy

Contract policies represent permitted and prohibited actions over a certain asset. These actions can be limited further

> by constraints (temporal or spatial) and duties (e.g. deletion of the data after 30 days). Further information is

Reviewer 2: ```suggestion
by constraints (temporal or spatial) and duties ("e.g. deletion of the data after 30 days"). Further information is
```

provided in a separate [section](Policies.md).

> ## Data Request

Reviewer 2: ```suggestion
## Data request
```

> Data Request is a representation which creation initialize the data [transfer process](#transfer-process) on the
> consumer side. It contains a reference of the to be transferred [Asset](#asset) and
> the [Contract (Agreement)](#contract-agreement).

Reviewer 2: ```suggestion
A `DataRequest` is sent from the consumer to the provider after a successful contract negotiation to initiate the data transfer. It references the [asset](#asset) and [contract](#contract-agreement) as well as information about the [data destination](#data-address).
```

## Transfer process

Representation of the data transfer. Runs through a state machine which defines certain states like provisioning,

> transferring, completed.

Reviewer 2: Analogous to the `ContractNegotiation` this object captures the current state of a data transfer. This process is inherently asynchronous, so the `TransferProcess` objects are stored in a backing data store (`TransferProcessStore`).

  contains a policy which is the contract policy.

* #### Contract Negotiation

  Is the process when an organization, represented with a connector as a consumer, asks for access to a certain asset
  and supplies all the requirements specified in the contract offer. Some states of the contract negotiation are the
  requesting of an asset, the offers asked in the negotiation, the approval or the declining of it.

* #### Contract Agreement

  agreed usage between data provider and data consumer after a contract negotiation. It points to a contract offer and
  can have expiry or cancellation date.

#### Policy:

[policy]: Contract policies represent permitted and prohibited actions over a certain asset. These actions can be
limited further by constraints (temporal or spatial) and duties (e.g. deletion of the data after 30 days). Further
information is provided in a separate [section](Policies.md).

#### Data Request:

Data Request is a representation which creation initialize the data [transfer process](#transfer-process) on the
consumer side. It contains a reference of the to be transferred [Asset](#asset) and
the [Contract (Agreement)](#contract-agreement).

> #### Transfer process:

Reviewer 3: I would decrease the headline level to just `##` (or `###` for the subpoints under contract)

Representation of the data transfer. Runs through a state machine which defines certain states like provisioning,
transferring, completed.

> ![domain_model](diagrams/domain-model.png)

Reviewer 3: Wouldn't it make sense to move this to the top of the file, so that people directly can get an overview?
